_source/_posts/2022-04-11-java-cli-device-grant.adoc

```diff
@@ -0,0 +1,412 @@
+flowchart TD
+    A[Start Login] -- POST to /v1/device/authorize --> B{Can open Browser?};
+    B -- Yes --> C[Open Browser to \n response.verification_uri_complete];
+    B -- No --> D{show QR Code?};
+    D -- Yes --> E[Create QR Code with \n response.verification_uri_complete];
+    D -- No --> F[Instruct user visit \n response.verification_uri and \n input response.user_code];
+    C --> G[Wait for user to \n login on other device];
+    E --> G;
+    F --> G;
+    G[sleep for response.interval] --> H[POST to /v1/token with \n response.device_code];
+    H --> I{Is Success};
+    I -- Yes --> J(User Is authenticated);
+    I -- No --> G;
+////
+
+NOTE: This example will work the same with any Identity Provider (IdP) that supports the flow, but you may need to tweak the URLs.
+
+== OAuth 2.0 from the command line
+
+You might not think of your terminal as an "input restricted device", but that still doesn't mean you should start typing a website's passwords; there are a few problems related to this:
+
+* You should ONLY enter a password into the website it was created for.
+* Your terminal doesn't support the same factors your browser does.
+* You may be ssh'd into another computer.
+
+NOTE: The Authorization Device Grant requires a user. If you are trying to call APIs from an automated process or other server-to-server communication, check out the link:/blog/2021/05/05/client-credentials-spring-security[OAuth 2.0 Client Credential Flow].
+
+== Setup Okta
+
+First, you need to create an Okta application that supports the device grant.
+
+. In **Applications**->**Applications**, click **Create App Integration**. In the pop-up, choose **OIDC - OpenID Connect**, then **Native Application**. The Device Grant is only available for Native Applications.
++
+image::{% asset_path 'blog/java-cli-device-grant/create-native-app.jpg' %}[alt=Set up a native app,width=800,align=center]
+
+. In the next step, check **Device Authorization** in the **Grant type** section.
++
+image::{% asset_path 'blog/java-cli-device-grant/enable-device-flow.jpg' %}[alt=Enable device grant,width=800,align=center]
+
+Make a note of the **Client ID**, you will need it below.
+
+If you are using an https://developer.okta.com/docs/concepts/auth-servers/#available-authorization-server-types[Okta Authorization Server], you also need to enable the grant type in the authorization server's policy interface:
+
+. Go to **Security**->**API**
+. Select the authorization server you want to use (for example, "default").
+. Click the **Access Policies** tab, and edit the rule for your policy. Make sure to select the **Device Authorization** check box.
++
+image::{% asset_path 'blog/java-cli-device-grant/custom-as.jpg' %}[alt=Enable device flow on a custom Authorization Server,width=800,align=center]
+
+Make note of your Issuer URI, if you are using an Okta Custom Authorization Server, it will look something like this (updating the hostname to match your Okta Org):
```

Comment:
```suggestion
Make note of your Issuer URI. If you are using an Okta Custom Authorization Server, it will look something like this (updating the hostname to match your Okta Org):
```
2 sentences